# Lab 4 - Information Theory in Machine Learning

Welcome to this week's lab on Information Theory! This week, we will dive into the fascinating world of Information Theory as applied to Machine Learning. Specifically, we will focus on two key concepts: Entropy and Information Gain. These principles are fundamental in understanding how decision trees make split decisions to organize data effectively.

### Entropy
- Entropy, in the context of information theory, measures the level of uncertainty or disorder within a set of data.
- In machine learning, particularly in decision trees, entropy helps to determine how a dataset should be split. A high entropy means more disorder, indicating that our dataset is varied. Conversely, low entropy suggests more uniformity in the data.

### Information Gain
- Information Gain measures the reduction in entropy after the dataset is split on an attribute.
- It is crucial in building decision trees as it helps to decide the order of attributes the tree will use for splitting the data. The attribute with the highest Information Gain is chosen as the splitting attribute at each node.

## Part 1: Entropy and Information Gain in Decision Trees
Decision Trees use these concepts to create branches. By choosing splits that maximize Information Gain (or equivalently minimize entropy), a decision tree can effectively categorize data, leading to better classification or regression models.

### Step 1: Import Necessary Libraries

In [1]:
import numpy as np
import pandas as pd
from sklearn.datasets import load_iris
from sklearn.tree import DecisionTreeClassifier
from sklearn import tree
import matplotlib.pyplot as plt

### Step 2: Load and Explore the Iris Dataset

In [3]:
iris = load_iris()
df = pd.DataFrame(iris.data, columns=iris.feature_names)
df['target'] = iris.target

### Step 3: Calculate Entropy
To calculate the `entropy` we need to:
- First, extract the target variable `y` from your dataset (like the 'target' column in the Iris dataset).
- Then, call `calculate_entropy(y)` to get the entropy.

This function calculates the entropy of a given target variable `y`. It works by first determining the unique classes in `y`, then computes the probability of each class, and uses this probability to calculate the entropy. This is a crucial step in understanding the disorder or uncertainty in the dataset, a fundamental concept in information theory.

In [4]:
def calculate_entropy(y):
    class_labels = np.unique(y)
    entropy = 0
    for label in class_labels:
        probability = len(y[y == label]) / len(y)
        entropy -= probability * np.log2(probability)
    return entropy

What is your observastion about the calculated Entropy?

This observation provides a insight about the level of impurity and disorder in a dataset.For example: if the entropy calculated for dataset was 0.981,then there is a room improving purity through splitting.

### Step 4: Calculate Information Gain
There are three steps for calculating the Information Gain:
1. Compute Overall Entropy: Use the entropy function from Step 3 on the entire target dataset.
2. Calculate Weighted Entropy for Each Attribute: For each unique value in the attribute, partition the dataset and calculate its entropy. Then calculate the weighted sum of these entropies, where the weights are the proportions of instances in each partition.
3. Compute Information Gain: Subtract the weighted entropy of the split from the original entropy.

The attribute with the highest Information Gain is generally chosen for splitting, as it provides the most significant reduction in uncertainty. This step is critical in constructing an effective decision tree, as it directly influences the structure and depth of the tree.

In [ ]:
def calculate_information_gain(df, attribute, target_name):
    total_entropy = calculate_entropy(df[target_name])
    values, counts = np.unique(df[attribute], return_counts=True)
    weighted_entropy = sum((counts[i] / sum(counts)) * calculate_entropy(df.where(df[attribute] == values[i]).dropna()[target_name]) for i in range(len(values)))
    information_gain = total_entropy - weighted_entropy
    return information_gain


Discuss your findings here.

## Part 2: Apply Entropy and Information Gain on a different dataset

Your task is to choose a new dataset and implement what you learned in `Part 1` on this new dataset.

### Task 1: Implement Entropy and Information Gain

In [5]:
# Your code goes here
#sample dataset
data = {
    'Weather': ['Sunny', 'Sunny', 'Rainy', 'Overcast', 'Rainy', 'Overcast', 'Sunny', 'Rainy', 'Overcast'],
    'Temperature': ['Hot', 'Mild', 'Cold', 'Mild', 'Mild', 'Hot', 'Cold', 'Hot', 'Cold'],
    'Play Sport': ['No', 'No', 'Yes', 'Yes', 'Yes', 'Yes', 'No', 'No', 'Yes']
}

df = pd.DataFrame(data)

# Define the entropy calculation
def calculate_entropy(y):
    class_labels = np.unique(y)
    entropy = 0
    for label in class_labels:
        probability = len(y[y == label]) / len(y)
        entropy -= probability * np.log2(probability)
    return entropy

# Calculate the total entropy of the target variable 'Play Sport'
total_entropy = calculate_entropy(df['Play Sport'])

# Define the information gain calculation function
def calculate_information_gain(df, attribute, target_name):
    total_entropy = calculate_entropy(df[target_name])
    values, counts = np.unique(df[attribute], return_counts=True)
    weighted_entropy = sum((counts[i] / sum(counts)) * calculate_entropy(df.where(df[attribute] == values[i]).dropna()[target_name]) for i in range(len(values)))
    information_gain = total_entropy - weighted_entropy
    return information_gain

# Calculate the information gain for the attribute 'Weather'
info_gain_weather = calculate_information_gain(df, 'Weather', 'Play Sport')

# Calculate the information gain for the attribute 'Temperature'
info_gain_temperature = calculate_information_gain(df, 'Temperature', 'Play Sport')

# Print the results
print("Total Entropy of 'Play Sport':", total_entropy)
print("Information Gain for 'Weather':", info_gain_weather)
print("Information Gain for 'Temperature':", info_gain_temperature)

Total Entropy of 'Play Sport': 0.9910760598382222
Information Gain for 'Weather': 0.6849774484867257
Information Gain for 'Temperature': 0.07278022578373267


### Task 2: Discuss your findings in detail
Provide detailed explanation and discussion about your findings.

The result from the entropy and information gain calculations show how well each attribute contributes to predict whether a person will play a sport or not. 

Total Entropy of Target Variable (Play Spor:t 

The total entropy of Play Sport was calculated as approximately 0.991. This value indicates the level of uncertainty or disorder in the target variable, Play Sport, without considering any other attributes. 

Information Gain for W:ether 

The information gain for Weather was calculated as approximately 0.685. Information gain (IG) measures how much entropy is reduced by splitting the dataset based on the chosen attribute. Higher information gain indicates a more significant reduction in entropy, meaning the attribute is more useful for classification. 

Information Gain for T:eperature 

The information gain for Temperature was calculated as approximately 0.073. This value indicates that Temperature contributes very little to reducing entropy, so it’s not as useful for classification in this dataset. :

Conclusion 

The analysis shows that Weather is the best attribute to use for predicting Play Sport due to its high information gain, meaning it has a substantial impact on reducing uncertainty. Temperature, on the other hand, has a low information gain, indicating it is not particularly helpful in this classification task. 

## Submission
Submit a link to your completed Jupyter Notebook file hosted on your private GitHub repository through the submission link in Blackboard.